primitives/state-machine/src/ext.rs

```diff
@@ -520,20 +548,10 @@ where
 								.flat_map(|(map, _)| map.clone().into_iter().map(|(k, v)| (k, v.value)))
 						);
 
-					self.backend.child_storage_root(storage_key, child_info.as_ref(), delta)
+					self.backend.child_storage_root(storage_key, child_info, delta)
 				};
 
 				let root = root.encode();
-				// We store update in the overlay in order to be able to use 'self.storage_transaction'
-				// cache. This is brittle as it rely on Ext only querying the trie backend for
-				// storage root.
-				// A better design would be to manage 'child_storage_transaction' in a
-				// similar way as 'storage_transaction' but for each child trie.
-				if is_empty {
```

Comment:
Note, this part of code is replaced by a caching in transaction_child_storage_root in storage_transaction_cache, this is more correct than the previous code that did rely on assuming root is always queried on the backend trie when accessing child content.